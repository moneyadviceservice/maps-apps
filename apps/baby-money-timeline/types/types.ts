export type Group = {
  title: string;
  group: string;
  descritionScoreOne?: string;
  descritionScoreTwo?: string;
  descritionScoreThree?: string;
};

export type Links = {
  title: string;
  link: string;
  type: string;
  description?: string;
};

export type FormContentAnlyticsData = {
  pageName: string;
  pageTitle: string;
  toolName: string;
  stepNames: string[] | string;
  categoryLevels?: string[];
};

export const FORM_FIELDS = {
  day: 'day',
  month: 'month',
  year: 'year',
  dueDateMin: 'dueDateMin',
  invalidDate: 'invalidDate',
} as const;

export type Fields = {
  [Key in keyof typeof FORM_FIELDS]: string;
};
