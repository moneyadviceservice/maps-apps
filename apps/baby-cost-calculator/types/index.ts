export type BabyCostTabIndex = 1 | 2 | 3 | 4 | 5;

export type Errors = Record<string, string[]> | null;
