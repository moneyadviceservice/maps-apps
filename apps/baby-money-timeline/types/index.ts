export type { FormContentAnlyticsData, Group, Links } from './types';
