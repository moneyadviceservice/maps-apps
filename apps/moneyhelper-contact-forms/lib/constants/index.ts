export enum FlowName {
  DEFAULT = 'money-management',
  PENSIONS_AND_RETIREMENT = 'pensions-and-retirement',
  DEBT_ADVICE = 'debt-advice',
  INSURANCE = 'insurance',
  SCAMS = 'scams',

  /**** PENSIONS_AND_RETIREMENT Flow ****/
  PENSIONS_STATE_PENSION = 'pensions-state-pension',
  PENSIONS_APPOINTMENTS = 'pensions-appointments',
  PENSIONS_TRACING = 'pensions-tracing',
  MHPD = 'mhpd',
  PENSIONS_GUIDANCE = 'pensions-guidance',

  /**** INSURANCE Flow ****/
  INSURANCE_TRAVEL = 'insurance-travel',
  INSURANCE_OTHER = 'insurance-other',

  /**** PENSIONS_APPOINTMENTS Flow  ****/
  APPOINTMENT_PENSION_WISE = 'appointment-pension-wise',
  APPOINTMENT_DIVORCE = 'appointment-divorce',
  // APPOINTMENT_SELF_EMPLOYED = 'appointment-self-employed',
}

export enum StepName {
  GUIDANCE = 'guidance',
  ENQUIRY_TYPE = 'enquiry-type',
  INSURANCE_TYPE = 'insurance-type',
  PENSION_TYPE = 'pension-type',
  APPOINTMENT_TYPE = 'appointment-type',
  NAME = 'name',
  DATE_OF_BIRTH = 'date-of-birth',
  CONTACT_DETAILS = 'contact-details',
  ENQUIRY = 'enquiry',
  CONFIRMATION = 'confirmation',
  ABOUT_SCAMS = 'about-scams',
  ABOUT_DEBT = 'about-debt',
  ABOUT_INSURANCE = 'about-insurance',
  ABOUT_STATE_PENSION = 'about-state-pension',
  ABOUT_PENSION_TRACING = 'about-pension-tracing',
  ABOUT_MHPD = 'about-mhpd',
  ABOUT_MONEY_MANAGEMENT = 'about-money-management',
  ABOUT_PENSION_WISE = 'about-pension-wise',
  ABOUT_PENSION_DIVORCE = 'about-pension-divorce',
  LOADING = 'loading',
  ERROR = 'error',
}

export enum Guards {
  COOKIE_GUARD = 'cookieGuard',
  VALIDATE_STEP_GUARD = 'validateStepGuard',
  AUTO_ADVANCE_GUARD = 'autoAdvanceGuard',
}

/**
 * Date of Birth validation constants.
 */
export const DOB = {
  day: { min: 1, max: 31 },
  month: { min: 1, max: 12 },
  year: {
    min: 1900,
    max: new Date().getFullYear(),
  },
};

/**
 * Error message lookup for various error codes returned by the API.
 * https://dev.azure.com/moneyandpensionsservice/MaPS%20Digital/_wiki/wikis/MaPS-Digital.wiki/1174/Azure-Function?anchor=response-to-react-webform-%22363974458063450487685701587858ae%22-%221e31ac0375494852a7fd5a9596257e04%22
 */
export const ERROR_MESSAGES: Record<string, string> = {
  '100':
    'Validation Failure – Invalid Payload. Your form submission is not valid.',
  '101':
    'Too Many Open Cases – Case Limit Reached for This Case Type. There was an issue in lodging your enquiry. Please contact us through another method.',
  '102':
    'Generic Failure – Contact/Case/CaseType Creation Issue in D365. Something went wrong while trying to create Case.',
  '103': 'Unknown error.',
  '104': 'Form submission failed. Please try again later.',
};

/**
 * A map for steps that require static headings regardless of the flow selected by the user. i.e. they are not dynamic based on the flow.
 */
export const HEADING_MAP: Record<string, string> = {
  [StepName.INSURANCE_TYPE]: StepName.INSURANCE_TYPE,
  [StepName.PENSION_TYPE]: StepName.PENSION_TYPE,
  [StepName.APPOINTMENT_TYPE]: StepName.APPOINTMENT_TYPE,
};

/**
 * List of flows that require a booking reference.
 * Used to determine if the booking reference field should be displayed on the enquiry step.
 */
export const FLOWS_WITH_BOOKING_REFERENCE = [
  FlowName.APPOINTMENT_PENSION_WISE,
  FlowName.APPOINTMENT_DIVORCE,
];

/**
 * List of flows that require a phone number.
 * Drives phone number validation (see isPhoneNumberRequired.ts), and used to drive UI logic so that the label does / doesn't include the word "optional".
 *
 */
export const FLOWS_WITH_REQUIRED_PHONE_NUMBER = [
  FlowName.APPOINTMENT_PENSION_WISE,
  FlowName.APPOINTMENT_DIVORCE,
];

/**
 * Mapping of flow names to enquiry types and kind of enquiry.
 * Required to form the payload.
 */
export const FLOW_TO_ENQUIRY_MAP: Record<
  string,
  { enquiryType: string; kindofEnquiry?: number }
> = {
  [FlowName.DEFAULT]: { enquiryType: 'money management' },
  [FlowName.PENSIONS_GUIDANCE]: { enquiryType: 'pensions - guidance' },
  [FlowName.DEBT_ADVICE]: { enquiryType: 'debt advice' },
  [FlowName.SCAMS]: { enquiryType: 'scams' },
  [FlowName.PENSIONS_STATE_PENSION]: {
    enquiryType: 'pensions - state pension',
    kindofEnquiry: 2,
  },
  [FlowName.PENSIONS_TRACING]: {
    enquiryType: 'pensions - appointments',
  },
  [FlowName.APPOINTMENT_PENSION_WISE]: {
    enquiryType: 'pensions - appointments',
    kindofEnquiry: 4,
  },
  [FlowName.APPOINTMENT_DIVORCE]: {
    enquiryType: 'pensions - appointments',
    kindofEnquiry: 5,
  },
  [FlowName.MHPD]: { enquiryType: 'MHPD', kindofEnquiry: 7 },
  [FlowName.INSURANCE_TRAVEL]: {
    enquiryType: 'insurance - travel',
  },
  [FlowName.INSURANCE_OTHER]: {
    enquiryType: 'insurance - other',
    kindofEnquiry: 8,
  },
};

/**
 * Mapping of flow names to step names for auto-advancing steps.
 * FlowName followed by the step name it should auto-advance to.
 */
export const AUTO_ADVANCE_STEP_MAP: Record<string, string> = {
  [FlowName.MHPD]: StepName.ABOUT_MHPD,
  // ... Add more mappings as needed
};
