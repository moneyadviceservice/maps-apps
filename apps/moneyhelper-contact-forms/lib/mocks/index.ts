export const mockRadioOptions = [
  {
    value: 'mock-value',
    text: 'mock-text',
  },
];
