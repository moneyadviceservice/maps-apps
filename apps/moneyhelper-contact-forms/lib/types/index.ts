export interface Entry {
  data: EntryData;
  stepIndex: number;
  errors: FormError;
}

export type EntryData = {
  flow: string;
  lang: string;
  [key: string]: string;
};

export type FormError = Record<string, (string | undefined)[]>; // Object with field names as keys and array of error messages as values

export type StepComponent = (props: {
  errors?: FormError;
  entry?: Entry;
  flow?: string;
  step: string;
  referenceNumber?: string;
}) => JSX.Element;

export type OptionTypesProps = (props: {
  errors?: FormError;
  step: string;
  optionsContentKey?: string;
  formErrorContentKey?: string;
}) => JSX.Element;

export type RouteConfig = {
  [key: string]: {
    Component: StepComponent;
    guards: string[];
    ui?: {
      [key: string]: boolean;
    };
  };
};

export type PageProps = {
  step: string;
  backStep: string;
  errors: FormError;
  flow: string;
  entry: Entry;
  referenceNumber?: string;
  url: string | undefined;
};
