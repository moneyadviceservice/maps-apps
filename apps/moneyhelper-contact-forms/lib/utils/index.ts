// export client-side safe utilities only
export * from './getCurrentStep';
export * from './getFieldError';
export * from './getFlowSteps';
export * from './getSessionId';
export * from './preparePayload';
export * from './safeT';
