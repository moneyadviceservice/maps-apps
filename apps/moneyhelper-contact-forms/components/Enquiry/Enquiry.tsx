import { Errors, ErrorType } from '@maps-react/common/components/Errors';
import { TextArea } from '@maps-react/form/components/TextArea';
import { TextInput } from '@maps-react/form/components/TextInput';
import useTranslation from '@maps-react/hooks/useTranslation';
import { Markdown } from '@maps-react/vendor/components/Markdown';

import { FlowName, FLOWS_WITH_BOOKING_REFERENCE } from '../../lib/constants';
import { StepComponent } from '../../lib/types';
import { getFieldError } from '../../lib/utils';
import { FormWrapper } from '../FormWrapper';

export const Enquiry: StepComponent = ({ errors, entry, flow, step }) => {
  const { t } = useTranslation();
  const showBookingReference = flow
    ? FLOWS_WITH_BOOKING_REFERENCE.includes(flow as FlowName)
    : false;
  return (
    <FormWrapper step={step} entry={entry} className="pt-4">
      <Errors
        errors={getFieldError('text-area', errors) ? [{} as ErrorType] : []}
      >
        <Markdown
          content={t(`components.enquiry.form.text-area.${flow}.label`)}
        />
        <TextArea
          id="text-area"
          name="text-area"
          error={
            getFieldError('text-area', errors)
              ? t('components.enquiry.form.text-area.error')
              : undefined
          }
          minLength={50}
          maxLength={4000}
          hasCharacterCounter
          defaultValue={entry?.data?.['text-area'] ?? ''}
          hasGlassBoxClass={true}
        />
      </Errors>
      {showBookingReference && (
        <div className="mt-14">
          <Errors
            errors={
              getFieldError('booking-reference', errors)
                ? [{} as ErrorType]
                : []
            }
          >
            <TextInput
              id="booking-reference"
              name="booking-reference"
              label={t('components.enquiry.form.booking-reference.label')}
              hint={t('components.enquiry.form.booking-reference.hint')}
              type="text"
              data-testid="input-booking-reference"
              error={
                getFieldError('booking-reference', errors)
                  ? t('components.enquiry.form.booking-reference.error')
                  : undefined
              }
              defaultValue={entry?.data?.['booking-reference'] ?? ''}
            />
          </Errors>
        </div>
      )}
    </FormWrapper>
  );
};
