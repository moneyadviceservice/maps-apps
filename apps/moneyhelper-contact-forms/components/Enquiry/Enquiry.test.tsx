import { render, screen } from '@testing-library/react';

import useTranslation from '@maps-react/hooks/useTranslation';

import { Entry, EntryData, FormError } from '../../lib/types';
import { Enquiry } from './Enquiry';

import '@testing-library/jest-dom';

jest.mock('@maps-react/hooks/useTranslation');
jest.mock('../../lib/constants', () => ({
  ...jest.requireActual('../../lib/constants'),
  FLOWS_WITH_BOOKING_REFERENCE: ['mock-flow'],
}));
jest.mock('../../lib/utils', () => ({
  ...jest.requireActual('../../lib/utils'),
  getFlowSteps: jest.fn().mockReturnValue(['mock-step']),
}));

const mockUseTranslation = useTranslation as jest.Mock;
const mockStep = 'mock-step';

describe('Enquiry Component', () => {
  beforeEach(() => {
    mockUseTranslation.mockReturnValue({
      t: (key: string) => key,
      z: (translations: { en: JSX.Element; cy: JSX.Element }) =>
        translations.en, // Always return the English translation for tests
    });
  });

  it('renders the component with title and form', () => {
    const { container } = render(<Enquiry step={mockStep} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with errors', () => {
    const mockErrors: FormError = { 'text-area': ['Description is required'] };

    const { container } = render(
      <Enquiry errors={mockErrors} step={mockStep} />,
    );

    expect(
      screen.getByText('components.enquiry.form.text-area.error'),
    ).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
  it('renders booking reference field when flow is in FLOWS_WITH_BOOKING_REFERENCE', () => {
    render(<Enquiry flow="mock-flow" step={mockStep} />);
    expect(
      screen.getByLabelText('components.enquiry.form.booking-reference.label'),
    ).toBeInTheDocument();
  });

  it('does not render booking reference field when flow is not in FLOWS_WITH_BOOKING_REFERENCE', () => {
    render(<Enquiry flow="not-in-list" step={mockStep} />);
    expect(
      screen.queryByLabelText(
        'components.enquiry.form.booking-reference.label',
      ),
    ).not.toBeInTheDocument();
  });

  it('renders booking reference with error', () => {
    const mockErrors: FormError = {
      'booking-reference': ['Booking reference is required'],
    };
    render(<Enquiry errors={mockErrors} flow="mock-flow" step={mockStep} />);
    expect(
      screen.getByText('components.enquiry.form.booking-reference.error'),
    ).toBeInTheDocument();
  });

  it('renders text area with default value', () => {
    const mockEntry = {
      data: {
        'text-area': 'Default text',
        flow: 'mock-flow',
        lang: 'en',
      } as EntryData,
    } as Entry;
    render(<Enquiry entry={mockEntry} step={mockStep} />);
    expect(screen.getByRole('textbox')).toHaveValue('Default text');
  });
});
