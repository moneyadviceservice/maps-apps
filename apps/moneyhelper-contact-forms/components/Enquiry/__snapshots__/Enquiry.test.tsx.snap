// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enquiry Component renders the component with title and form 1`] = `
<form
  action="/api/form-handler"
  aria-label="Contact form for Step: mock-step"
  class="pt-4"
  method="POST"
  novalidate=""
>
  <input
    name="lang"
    type="hidden"
    value=""
  />
  <div
    class=""
    data-testid="errors"
  >
    <p
      class="mb-4"
      data-testid="paragraph"
    >
      components.enquiry.form.text-area.undefined.label
    </p>
    <textarea
      aria-invalid="false"
      aria-required="true"
      class="obfuscate px-3 m-px w-full rounded border resize focus:outline-none focus:shadow-focus-outline tool-field min-h-[312px] focus:border-1 focus:border-blue-700 border-gray-400"
      id="text-area"
      maxlength="4000"
      minlength="50"
      name="text-area"
    />
    <p
      aria-live="polite"
      class="text-base text-gray-400"
    >
      You have 4,000 characters remaining.
    </p>
  </div>
  <button
    class="border-0 text-white bg-magenta-500 hover:bg-pink-800 active:bg-pink-900 active:outline-yellow-400 active:text-white t-button relative px-3 py-2 font-semibold rounded outline-none shadow-bottom-gray inline-flex items-center justify-between gap-2 cursor-pointer disabled:cursor-not-allowed disabled:border-slate-400 disabled:border disabled:bg-transparent disabled:text-gray-400 disabled:shadow-none focus:bg-yellow-400 focus:outline-blue-700 focus:outline-[3px] focus:outline-offset-0 focus:text-gray-800 focus:shadow-none focus:before:content-[""] focus:before:w-[calc(100%+6px)] focus:before:rounded-[0.45rem] focus:before:absolute focus:before:inset-0 focus:before:inset-x-[-3px] focus:before:inset-y-[-3px] focus:before:shadow-bottom-gray active:outline-[3px] active:outline-offset-0 active:shadow-inset active:before:content-[""] active:before:w-[calc(100%+6px)] active:before:rounded-[0.45rem] active:before:absolute active:before:inset-0 active:before:inset-x-[-3px] active:before:inset-y-[-3px] active:before:shadow-bottom-white t-primary-button w-full mt-10 mb-4 md:w-auto"
    data-testid="continue-button"
    type="submit"
  >
    <span />
    common.form.button.continue-text
    <span />
  </button>
</form>
`;

exports[`Enquiry Component renders with errors 1`] = `
<form
  action="/api/form-handler"
  aria-label="Contact form for Step: mock-step"
  class="pt-4"
  method="POST"
  novalidate=""
>
  <input
    name="lang"
    type="hidden"
    value=""
  />
  <div
    class="border-0 border-l-4 pl-5 border-red-600 border-solid"
    data-testid="errors"
  >
    <p
      class="mb-4"
      data-testid="paragraph"
    >
      components.enquiry.form.text-area.undefined.label
    </p>
    <p
      class="mb-4 text-red-700"
      data-testid="text-area-error"
      id="text-area-error"
    >
      components.enquiry.form.text-area.error
    </p>
    <textarea
      aria-errormessage="text-area-error"
      aria-invalid="true"
      aria-required="true"
      class="obfuscate px-3 m-px w-full rounded resize focus:outline-none focus:shadow-focus-outline tool-field min-h-[312px] focus:border-1 focus:border-blue-700 border-red-700 border-2"
      id="text-area"
      maxlength="4000"
      minlength="50"
      name="text-area"
    />
    <p
      aria-live="polite"
      class="text-base text-gray-400"
    >
      You have 4,000 characters remaining.
    </p>
  </div>
  <button
    class="border-0 text-white bg-magenta-500 hover:bg-pink-800 active:bg-pink-900 active:outline-yellow-400 active:text-white t-button relative px-3 py-2 font-semibold rounded outline-none shadow-bottom-gray inline-flex items-center justify-between gap-2 cursor-pointer disabled:cursor-not-allowed disabled:border-slate-400 disabled:border disabled:bg-transparent disabled:text-gray-400 disabled:shadow-none focus:bg-yellow-400 focus:outline-blue-700 focus:outline-[3px] focus:outline-offset-0 focus:text-gray-800 focus:shadow-none focus:before:content-[""] focus:before:w-[calc(100%+6px)] focus:before:rounded-[0.45rem] focus:before:absolute focus:before:inset-0 focus:before:inset-x-[-3px] focus:before:inset-y-[-3px] focus:before:shadow-bottom-gray active:outline-[3px] active:outline-offset-0 active:shadow-inset active:before:content-[""] active:before:w-[calc(100%+6px)] active:before:rounded-[0.45rem] active:before:absolute active:before:inset-0 active:before:inset-x-[-3px] active:before:inset-y-[-3px] active:before:shadow-bottom-white t-primary-button w-full mt-10 mb-4 md:w-auto"
    data-testid="continue-button"
    type="submit"
  >
    <span />
    common.form.button.continue-text
    <span />
  </button>
</form>
`;
